src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1393,29 +1333,249 @@ RewriteResponse TheoryStringsRewriter::preRewrite(TNode node) {
+      return returnRewrite(node, ret, "ss-len-include");
+    }
+  }
+
+  // symbolic length analysis
+  for (unsigned r = 0; r < 2; r++)
+  {
+    // the amount of characters we can strip
+    Node curr;
+    if (r == 0)
+    {
+      if (node[1] != zero)
+      {
+        // strip up to start point off the start of the string
+        curr = node[1];
+      }
+    }
+    else if (r == 1)
+    {
+      Node tot_len = Rewriter::rewrite(
+          NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, node[0]));
+      Node end_pt = Rewriter::rewrite(
+          NodeManager::currentNM()->mkNode(kind::PLUS, node[1], node[2]));
+      if (node[2] != tot_len)
+      {
+        if (checkEntailArith(node[2], tot_len))
+        {
+          // end point beyond end point of string, map to tot_len
+          Node ret = NodeManager::currentNM()->mkNode(
+              kind::STRING_SUBSTR, node[0], node[1], tot_len);
+          return returnRewrite(node, ret, "ss-end-pt-norm");
+        }
+        else
+        {
+          // strip up to ( str.len(node[0]) - end_pt ) off the end of the string
+          curr = Rewriter::rewrite(
+              NodeManager::currentNM()->mkNode(kind::MINUS, tot_len, end_pt));
+        }
+      }
+    }
+    if (!curr.isNull())
+    {
+      // strip off components while quantity is entailed positive
+      int dir = r == 0 ? 1 : -1;
+      std::vector<Node> childrenr;
+      if (stripSymbolicLength(n1, childrenr, dir, curr))
+      {
+        if (r == 0)
+        {
+          Node ret = NodeManager::currentNM()->mkNode(
+              kind::STRING_SUBSTR,
+              mkConcat(kind::STRING_CONCAT, n1),
+              curr,
+              node[2]);
+          return returnRewrite(node, ret, "ss-strip-start-pt");
+        }
+        else
+        {
+          Node ret = NodeManager::currentNM()->mkNode(
+              kind::STRING_SUBSTR,
+              mkConcat(kind::STRING_CONCAT, n1),
+              node[1],
+              node[2]);
+          return returnRewrite(node, ret, "ss-strip-end-pt");
+        }
+      }
+    }
+  }
+  // combine substr
+  if (node[0].getKind() == kind::STRING_SUBSTR)
+  {
+    Node start_inner = node[0][1];
+    Node start_outer = node[1];
+    if (checkEntailArith(start_outer) && checkEntailArith(start_inner))
+    {
+      // both are positive
+      // thus, start point is definitely start_inner+start_outer.
+      // We can rewrite if it for certain what the length is
+
+      // the length of a string from the inner substr subtracts the start point
+      // of the outer substr
+      Node len_from_inner = Rewriter::rewrite(NodeManager::currentNM()->mkNode(
+          kind::MINUS, node[0][2], start_outer));
+      Node len_from_outer = node[2];
+      Node new_len;
+      // take quantity that is for sure smaller than the other
+      if (len_from_inner == len_from_outer)
+      {
+        new_len = len_from_inner;
+      }
+      else if (checkEntailArith(len_from_inner, len_from_outer))
```

Comment:
We could use under-/overapproximations to enable additional rewrites here. See comment on line 2045 below.

Reply:
Yes, these should be internal to checkEntailArith (line 2544)